In [1]:
import numpy as np
from sklearn.datasets import load_diabetes
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

In [2]:
X,y=load_diabetes(return_X_y=True)

In [3]:
X.shape

(442, 10)

In [4]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=2)

In [5]:
X_train.shape

(309, 10)

In [14]:
X_train.shape[1]

10

In [7]:
lr= LinearRegression()

In [8]:
lr.fit(X_train,y_train)

LinearRegression()

In [9]:
lr.coef_

array([ -31.80406415, -214.66452697,  445.14791885,  376.34786891,
       -802.50380919,  525.99207358,  106.4287134 ,  172.3171033 ,
        799.50335529,   57.1646797 ])

In [10]:
lr.intercept_

152.6275795791894

In [11]:
y_pred=lr.predict(X_test)

In [12]:
r2_score(y_test,y_pred)

0.5012577617513011

### self mini-batch gradient descent

In [13]:
import random

In [55]:
class myMBGD:

    def __init__(self,batch_size,learning_rate=0.01,epochs=50):
        self.coef_=None
        self.intercept_=None
        self.lr=learning_rate
        self.epochs=epochs
        self.batch_size=batch_size
       
    def fit(self,X_train,y_train):
        self.intercept_=0 #initialize slope with 0
        self.coef_=np.ones(X_train.shape[1]) # initialize the betas with one


        for i in range(self.epochs):
            
            for j in range(int(X_train.shape[0]/self.batch_size)): #divide data into desired batches

                ranx = random.sample(range(X_train.shape[0]),self.batch_size)  # randomly choose batch features
                    
            
                y_hat = np.dot(X_train[ranx],self.coef_) + self.intercept_ #calculate yhat for the random batch until loop run
                intercept_der=-2* np.mean(y_train[ranx]-y_hat) 
                # calculate intercept for batch
                
                self.intercept_= self.intercept_ - (self.lr*intercept_der)

            
                coef_der=-2* np.dot((y_train[ranx]-y_hat),X_train[ranx])

                #calculate coefficients
                self.coef_=self.coef_ -(self.lr*coef_der)
            
        print(self.intercept_,self.coef_)


    def predict(self,X_test):
        return np.dot(X_test,self.coef_)+self.intercept_






In [84]:
mbr = myMBGD(batch_size=int(X_train.shape[0]/10),learning_rate=0.01,epochs=170)

In [85]:
mbr.fit(X_train,y_train)

151.4191267625855 [  -1.95348809 -181.25046346  414.08925119  337.28981756  -20.41875451
  -90.967621   -212.08831484  133.84755549  443.04078959   99.59787941]


In [86]:
y_pred=mbr.predict(X_test)

In [87]:
r2_score(y_test,y_pred)

0.49429389498440324

### compare with sgd mini batch

In [151]:
from sklearn.linear_model import SGDRegressor

In [180]:
sgd=SGDRegressor(learning_rate='constant',eta0=0.09)

In [181]:
#use partial_fit to divide data into batches

In [182]:
batch_size = 35
for i in range(100):
    ranx=random.sample(range(X_train.shape[0]),batch_size)
    sgd.partial_fit(X_train[ranx],y_train[ranx])

In [183]:
sgd.coef_

array([  44.36427804,  -67.59873456,  293.80614018,  225.98318832,
         33.65793946,   -7.54581433, -180.84416748,  148.93345455,
        315.41389959,  125.01113738])

In [184]:
sgd.intercept_

array([146.35176902])

In [185]:
y_pred=sgd.predict(X_test)

In [186]:
from sklearn.metrics import r2_score

r2 = r2_score(y_test, y_pred)
print("R² Score:", r2)


R² Score: 0.43533773819236476


### observation

linear regression r2_score 0.50

self gd r2_score 0.49

sklearn sgd r2_score 0.43

using stochastic gd ->

self sto_gd r2_score 0.42

sklearn sgd r2_score 0.40

as stochastic gd is used for large dataset, thus in this case mini batch performed better